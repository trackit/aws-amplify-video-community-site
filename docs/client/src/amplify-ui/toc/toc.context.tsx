import {createProviderConsumer} from "@stencil/state-tunnel";
import {h} from "@stencil/core";
import {TOCContext} from "./toc.types";

export const tocContext = createProviderConsumer<TOCContext>(
  {},
  (subscribe, child) => <context-consumer {...{subscribe}} renderer={child} />,
);
