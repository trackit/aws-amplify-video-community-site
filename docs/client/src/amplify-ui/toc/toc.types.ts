export type SetContent = (content: HTMLElement) => void;

export interface TOCContext {
  elements?: HTMLElement[];
  setContent?: SetContent;
}
