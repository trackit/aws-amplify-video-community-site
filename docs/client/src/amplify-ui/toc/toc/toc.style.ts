import {css} from "emotion";
import {MQTablet, MQDesktop} from "../../styles/media";

export const tocStyle = css`
  display: none;
  flex-direction: column;
  padding: 2rem 0;
  font-size: 0.875rem;
  width: 16.875rem;

  ${MQTablet} {
    &.more-width {
      display: flex;
    }
  }

  ${MQDesktop} {
    display: flex;
  }

  docs-in-page-link {
    transition: 0.25s ease;
    transition-property: margin-left border;
    border-left: 0.05rem solid var(--border-color);
    display: flex;

    a {
      display: flex;
      flex: 1;
      padding-top: 0.25rem;
      padding-bottom: 0.25rem;
      padding-right: 1.5rem;
      color: var(--font-color);
    }

    &:hover {
      background-color: var(--bg-color-hover);
    }

    &:first-child {
      margin-top: 0;
    }

    &.active {
      border-left-color: var(--color-anchor);

      a {
        color: var(--color-anchor);
      }
    }
  }
`;

export const h2AnchorStyle = css`
  a {
    padding-left: 1.5rem;
  }
`;

export const h3AnchorStyle = css`
  a {
    padding-left: 2rem !important;
    font-size: 0.8125rem;
    line-height: 1.25rem;
  }
`;

export const headerStyle = css`
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-bottom: 0.5rem;
  border-left: 0.05rem solid var(--border-color);
`;
