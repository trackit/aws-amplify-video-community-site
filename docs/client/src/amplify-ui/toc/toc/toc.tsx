import {Component, Host, Listen, Prop, State, h} from "@stencil/core";
import {tocContext} from "../toc.context";
import {sidebarLayoutContext} from "../../sidebar-layout/sidebar-layout.context";
import {tocStyle, h2AnchorStyle, h3AnchorStyle, headerStyle} from "./toc.style";
import {getElementTop} from "../../../utils/get-element-top";

const headingStyleByTagName = {
  H2: h2AnchorStyle,
  H3: h3AnchorStyle,
};

@Component({tag: "amplify-toc", shadow: false})
export class AmplifyTOC {
  /**
   * A list of `h2` and/or `h3` nodes, provided by either of 2 means:
   * 1. User-provided (`<amplify-toc elements={[...h2h3DomNodes]} />`)
   * 2. Provider-injected (within parent `amplify-toc-provider`, sibling
   * an `amplify-toc-content` instance)
   */
  @Prop() readonly elements?: HTMLElement[];
  /**
   * Whether or not the left sidebar is in view; injected from the sidebar context.
   * If the left sidebar isn't in view, there is more space for this table of contents
   */
  @Prop() readonly inView?: boolean;
  /*** the title of the page on which this TOC is being rendered */
  @Prop() readonly pageTitle?: string;
  /*** offset the active item (useful when combatting sticky header) */
  @Prop() readonly stickyHeaderHeight = 54;

  @State() activeLinkI?: number;
  previous?: number;

  @Listen("scroll", {target: "window"})
  @Listen("resize", {target: "window"})
  setActiveLink() {
    if (this.elements) {
      let i = this.elements.findIndex(
        (e) => getElementTop(e, this.stickyHeaderHeight) - 3 > window.scrollY,
      );
      if (i === -1) {
        i = this.elements.length;
      }
      this.activeLinkI = i - 1;
      if (this.activeLinkI !== this.previous) {
        this.previous = this.activeLinkI;
        if (this.activeLinkI >= 0) {
          const activeElement = this.elements[this.activeLinkI];
          if (activeElement) {
            history.replaceState(
              undefined,
              document.title,
              `#${activeElement.id}`,
            );
          }
        } else {
          history.replaceState(
            undefined,
            document.title,
            window.location.href.split("#")[0],
          );
        }
      }
    }
  }

  render() {
    return (
      this.elements &&
      this.elements.length > 0 && (
        <Host class={(this.inView ? "" : "more-width ") + tocStyle}>
          <div>
            <h4 class={headerStyle}>{this.pageTitle || "Contents"}</h4>
            {this.elements.map((e, i) => {
              const headingAnchorClass = headingStyleByTagName[e.tagName];
              return (
                <docs-in-page-link
                  targetId={e.id}
                  class={{
                    active: i === this.activeLinkI,
                    [headingAnchorClass]: true,
                  }}
                >
                  <div innerHTML={e.innerHTML} />
                </docs-in-page-link>
              );
            })}
          </div>
        </Host>
      )
    );
  }
}

tocContext.injectProps(AmplifyTOC, ["elements"]);
sidebarLayoutContext.injectProps(AmplifyTOC, ["inView"]);
