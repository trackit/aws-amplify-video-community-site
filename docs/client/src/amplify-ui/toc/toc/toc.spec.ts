import {AmplifyTOC} from "./toc";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-toc", () => {
  describe("Component logic", () => {
    let toc: AmplifyTOC;
    beforeEach(() => (toc = new AmplifyTOC()));

    it("should init `elements` as `undefined`", () =>
      expect(toc.elements).toBeUndefined());

    it("should init `activeLinkI` as `undefined`", () =>
      expect(toc.activeLinkI).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyTOC],
            html: `<amplify-toc />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
