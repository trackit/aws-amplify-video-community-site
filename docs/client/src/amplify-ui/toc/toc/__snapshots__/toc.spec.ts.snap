// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-toc Render logic should render 1`] = `<amplify-toc></amplify-toc>`;
